File: clients/go/codec_test.go
package fibril

import (
	"bytes"
	"encoding/hex"
	"reflect"
	"testing"
)

func TestFrameHeaderByteLayout(t *testing.T) {
	// op=publishOk(25=0x19), flags=0, request_id=777(0x309), payload=[0xaa,0xbb].
	f := buildFrame(opPublishOk, 777, []byte{0xaa, 0xbb})
	got := hex.EncodeToString(encodeFrame(f))
	// u32 len=2, u16 ver=1, u16 op=0x0019, u32 flags=0, u64 reqid=0x0309, payload.
	want := "00000002" + "0001" + "0019" + "00000000" + "0000000000000309" + "aabb"
	if got != want {
		t.Errorf("frame header layout\n got: %s\nwant: %s", got, want)
	}
}

func TestFrameRoundTrip(t *testing.T) {
	orig := buildFrame(opDeliver, 0xdeadbeef, []byte("hello body"))
	dec, consumed, ok := tryDecodeFrame(encodeFrame(orig))
	if !ok {
		t.Fatal("tryDecodeFrame reported incomplete on a whole frame")
	}
	if consumed != frameHeaderSize+len(orig.Payload) {
		t.Errorf("consumed = %d, want %d", consumed, frameHeaderSize+len(orig.Payload))
	}
	if !reflect.DeepEqual(dec, orig) {
		t.Errorf("round-trip mismatch\n got: %+v\nwant: %+v", dec, orig)
	}
}

func TestTryDecodeFramePartialAndMulti(t *testing.T) {
	a := encodeFrame(buildFrame(opPing, 1, nil))
	b := encodeFrame(buildFrame(opPong, 2, []byte{9}))

	// A header-length-minus-one prefix is incomplete.
	if _, _, ok := tryDecodeFrame(a[:frameHeaderSize-1]); ok {
		t.Error("expected incomplete on a short header")
	}
	// A full body minus one byte is incomplete.
	if _, _, ok := tryDecodeFrame(b[:len(b)-1]); ok {
		t.Error("expected incomplete on a truncated body")
	}

	// Two frames back to back decode one at a time from the running buffer.
	buf := append(append([]byte{}, a...), b...)
	f1, n1, ok := tryDecodeFrame(buf)
	if !ok || f1.Opcode != opPing {
		t.Fatalf("first frame decode failed: ok=%v op=%d", ok, f1.Opcode)
	}
	buf = buf[n1:]
	f2, n2, ok := tryDecodeFrame(buf)
	if !ok || f2.Opcode != opPong || !bytes.Equal(f2.Payload, []byte{9}) {
		t.Fatalf("second frame decode failed: ok=%v op=%d payload=%v", ok, f2.Opcode, f2.Payload)
	}
	if len(buf[n2:]) != 0 {
		t.Errorf("expected buffer fully consumed, %d bytes left", len(buf[n2:]))
	}
}

File: clients/csharp/Fibril.Tests/CodecTest.cs
using System.Text;
using Fibril;

namespace Fibril.Tests;

public class CodecTest
{
    private static string Hex(byte[] b) => Convert.ToHexString(b).ToLowerInvariant();

    [Fact]
    public void FrameHeaderByteLayout()
    {
        // op=PublishOk(25=0x19), flags=0, request_id=777(0x309), payload=[0xaa,0xbb].
        var f = Frame.Build(Op.PublishOk, 777, new byte[] { 0xaa, 0xbb });
        var got = Hex(f.Encode());
        // u32 len=2, u16 ver=1, u16 op=0x0019, u32 flags=0, u64 reqid=0x0309, payload.
        var want = "00000002" + "0001" + "0019" + "00000000" + "0000000000000309" + "aabb";
        Assert.Equal(want, got);
    }

    [Fact]
    public void FrameRoundTrip()
    {
        var orig = Frame.Build(Op.Deliver, 0xdeadbeef, Encoding.ASCII.GetBytes("hello body"));
        Assert.True(Frame.TryDecode(orig.Encode(), out var dec, out var consumed));
        Assert.Equal(Frame.HeaderSize + orig.Payload.Length, consumed);
        Assert.Equal(orig.Opcode, dec.Opcode);
        Assert.Equal(orig.RequestId, dec.RequestId);
        Assert.Equal(orig.Payload, dec.Payload);
    }

    [Fact]
    public void TryDecodePartialAndMulti()
    {
        var a = Frame.Build(Op.Ping, 1, null).Encode();
        var b = Frame.Build(Op.Pong, 2, new byte[] { 9 }).Encode();

        // A header-length-minus-one prefix is incomplete.
        Assert.False(Frame.TryDecode(a.AsSpan(0, Frame.HeaderSize - 1), out _, out _));
        // A full body minus one byte is incomplete.
        Assert.False(Frame.TryDecode(b.AsSpan(0, b.Length - 1), out _, out _));

        // Two frames back to back decode one at a time from the running buffer.
        var buf = new byte[a.Length + b.Length];
        a.CopyTo(buf, 0);
        b.CopyTo(buf, a.Length);
        Assert.True(Frame.TryDecode(buf, out var f1, out var n1));
        Assert.Equal(Op.Ping, f1.Opcode);
        Assert.True(Frame.TryDecode(buf.AsSpan(n1), out var f2, out var n2));
        Assert.Equal(Op.Pong, f2.Opcode);
        Assert.Equal(new byte[] { 9 }, f2.Payload);
        Assert.Equal(0, buf.Length - n1 - n2);
    }

    [Fact]
    public void Fnv1aCanonical()
    {
        var cases = new Dictionary<string, ulong>
        {
            [""] = 14695981039346656037,
            ["a"] = 12638187200555641996,
            ["order-42"] = 9015620992513762004,
            ["partition-key"] = 11792757095719117019,
            ["hello world"] = 8618312879776256743,
        };
        foreach (var (key, want) in cases)
        {
            Assert.Equal(want, Fnv.Hash1a(Encoding.UTF8.GetBytes(key)));
        }
    }

    [Fact]
    public void WireReaderRejectsTrailingBytes()
    {
        var r = new WireReader(new byte[] { 1, 2, 3 });
        Assert.Equal(1, r.U8());
        var ex = Assert.Throws<WireException>(() => r.Finish());
        Assert.Equal(WireErrorKind.TrailingBytes, ex.Kind);
    }

    [Fact]
    public void WireReaderRejectsShortRead()
    {
        var r = new WireReader(new byte[] { 0, 0 });
        var ex = Assert.Throws<WireException>(() => r.U32());
        Assert.Equal(WireErrorKind.UnexpectedEof, ex.Kind);
    }
}
